src/Traits/MockTrait.php

```diff
@@ -129,30 +130,64 @@ protected function assertArguments(
         string $class,
         string $function,
         int $callIndex,
-        bool $isClass = true
+        bool $isClass = true,
     ): void {
+        $reflection = ($isClass)
+            ? new ReflectionMethod($class, $function)
+            : new ReflectionFunction($function);
+
+        $reflectionArgs = $reflection->getParameters();
+
+        $this->assertArgumentsCount($actual, $expected, $reflectionArgs, $function);
+
+        $this->fillOptionalArguments($reflectionArgs, $actual, $expected, $isClass);
+
+        $message = ($isClass)
+            ? "Class '{$class}'\nMethod: '{$function}'\nMethod call index: {$callIndex}"
+            : "Namespace '{$class}'\nFunction: '{$function}'\nCall index: {$callIndex}";
+
+        $this->compareArguments($actual, $expected, $message);
+    }
+
+    protected function assertArgumentsCount(array $actual, array $expected, array $reflectionArgs, string $function): void
+    {
         $expectedCount = count($expected);
         $actualCount = count($actual);
+        $requiredParametersCount = count(array_filter($reflectionArgs, fn ($param) => !$param->isOptional()));
 
         if ($expectedCount !== $actualCount) {
-            $requiredParametersCount = count(array_filter($actual, fn ($item) => $item !== self::OPTIONAL_ARGUMENT_NAME));
+            $this->assertFalse(
+                $expectedCount < $requiredParametersCount,
+                "Failed assert that function {$function} was called with {$expectedCount} require arguments, actually it calls with {$requiredParametersCount} require arguments."
+            );
 
-            if ($expectedCount > $actualCount || $expectedCount < $requiredParametersCount) {
-                throw new Exception("Failed assert that function {$function} was called with {$expectedCount} arguments, actually it calls with {$actualCount} arguments.");
-            }
+            $this->assertFalse(
+                $expectedCount > $actualCount,
+                "Failed assert that function {$function} was called with {$expectedCount} arguments, actually it calls with {$actualCount} arguments."
+            );
         }
+    }
 
-        $expected = array_pad($expected, $actualCount, self::OPTIONAL_ARGUMENT_NAME);
+    protected function fillOptionalArguments(array $parameters, array &$actual, array &$expected, bool $isClass): void
+    {
+        foreach ($parameters as $index => $parameter) {
+            if (!isset($expected[$index]) && $parameter->isOptional()) {
+                $expected[$index] = $parameter->getDefaultValue();
+            }
 
-        $message = ($isClass)
-            ? "Class '{$class}'\nMethod: '{$function}'\nMethod call index: {$callIndex}"
-            : "Namespace '{$class}'\nFunction: '{$function}'\nCall index: {$callIndex}";
+            if (!$isClass && $actual[$index] === 'optionalParameter') {
+                $actual[$index] = $expected[$index] ?? $parameter->getDefaultValue();
```

Comment:
please move actual args filling before the expected args filling

```suggestion
                $actual[$index] = $parameter->getDefaultValue();
```